src/core/document.js

```diff
@@ -910,14 +864,52 @@ class PDFDocument {
       if (!(data instanceof BaseStream) || data.isEmpty) {
         continue;
       }
+      entries[name] = data;
+    }
+    return shadow(this, "xfaStreams", entries);
+  }
+
+  get xfaDatasets() {
+    const streams = this.xfaStreams;
+    if (!streams) {
+      return shadow(this, "xfaDatasets", null);
+    }
+    for (const key of ["datasets", "xdp:xdp"]) {
+      const stream = streams[key];
+      if (!stream) {
+        continue;
+      }
+      try {
+        const str = stringToUTF8String(stream.getString());
+        const data = Object.fromEntries([[key, str]]);
+        return shadow(this, "xfaDatasets", new DatasetReader(data));
+      } catch (_) {
+        warn("XFA - Invalid utf-8 string.");
+        break;
+      }
+    }
+    return shadow(this, "xfaDatasets", null);
+  }
+
+  get xfaData() {
```

Comment:
In the existing code we're not using `shadow` for this method, since it's only being accessed *once* from the `xfaFactory` getter and shadowing is being used there instead.